sdk/helper/storagepacker/invalidate.go

```diff
@@ -0,0 +1,295 @@
+	deleted := make([]*Item, 0)
+	revisit := make([]*Item, 0)
+
+	// Using the radix lookup will find child buckets, but not parent
+	// buckets.  For that we'll have to walk up the tree a step at a time.
+	//
+	// For example, we might be processing 00/0
+	//
+	// 00 -- 00/0 -- 00/0/a
+	//
+	// but, the item with key 000a... might exist in 00/0/a or maybe even in 00,
+	// so we want to check both.
+	//
+	// If we find it somewhere, return the version in the bucket with the longest key.
+	// (That is more likely to be a more recent version.)
+	partition, err := s.partitionRequests(maybeDeleted)
+	if err != nil {
+		return nil, nil, errwrap.Wrapf("problem searching for deleted keys: {{err}}", err)
+	}
+
+	// Looking through the other buckets without acquiring a lock
+	// is unsafe, but should be OK in the context of a replica.
+	for _, p := range partition {
+		bucketsToCheck := s.bucketAndAllParents(p.Bucket)
+		for _, request := range p.Requests {
+			found := false
+			for _, b := range bucketsToCheck {
+				// Skip originalBucket
+				if b.Key != bucketKey {
+					var data []byte
+					if data, found = b.ItemMap[request.ID]; found {
+						revisit = append(revisit, &Item{
+							ID:   request.ID,
+							Data: data,
+						})
+						break
+					}
+				}
+			}
+			if !found {
+				// Include the original value
+				// though maybe this is overkill and we should just return IDs.
+				deleted = append(deleted, request.Value)
+			}
+		}
+	}
+
+	return deleted, revisit, nil
+
+}
+
+// Assemble a list (from longest key to shortest) of all the parent buckets.
+func (s *StoragePackerV2) bucketAndAllParents(bucket *LockedBucket) []*LockedBucket {
+	s.bucketsCacheLock.RLock()
+	defer s.bucketsCacheLock.RUnlock()
+
+	parents := []*LockedBucket{bucket}
+	for k := s.parentCacheKey(bucket.Key); k != ""; k = s.parentCacheKey(k) {
+		bucketRaw, found := s.bucketsCache.Get(k)
+		if found {
+			parents = append(parents, bucketRaw.(*LockedBucket))
+		}
+	}
+	return parents
+}
+
+// Given a bucket key, return the cache key for its parent
+func (s *StoragePackerV2) parentCacheKey(bucketKey string) string {
+	bucketCacheKey := s.GetCacheKey(bucketKey)
+	if len(bucketCacheKey) < s.BaseBucketBits/4 {
+		return ""
+	} else {
+		n := len(bucketCacheKey)
+		shardLen := s.BucketShardBits / 4
+		return bucketCacheKey[:(n - shardLen)]
+	}
+}
+
+// Create a placeholder empty bucket given its key
+func (s *StoragePackerV2) newEmptyBucket(bucketKey string) *LockedBucket {
+	cacheKey := s.GetCacheKey(bucketKey)
+	lock := locksutil.LockForKey(s.storageLocks, cacheKey)
+	return &LockedBucket{
+		Bucket: &Bucket{
+			Key:       bucketKey,
+			HasShards: false,
+			ItemMap:   make(map[string][]byte),
+		},
+		LockEntry: lock,
+	}
+}
+
+func itemDifferenceBetween(originalBucket *LockedBucket, replacementBucket *LockedBucket) []*itemRequest {
+	maybeDeleted := make([]*itemRequest, 0)
```

Comment:
Curious what the benefit is here of doing `make([]*itemRequest, 0)`, versus `var maybeDeleted []*itemRequest`. I'm not committed to doing it one way or the other, I just have never actually thought of doing this before! :-)

Reply:
I think this is just me being a Go newbie and preferring always using the same way to create a slice. I didn't think there should be any difference, but it looks like I'm wrong.

I ran two versions through compiler explorer: https://godbolt.org/z/7ejxA0

The make() call doesn't get optimized away, so the "var" version would save the function call.